substrate/client/consensus/beefy/src/communication/request_response/outgoing_requests_engine.rs

```diff
@@ -98,12 +98,17 @@ impl<B: Block, AuthorityId: AuthorityIdBound> OnDemandJustificationsEngine<B, Au
 		}
 	}
 
-	fn reset_peers_cache_for_block(&mut self, block: NumberFor<B>) {
-		self.peers_cache = self.live_peers.lock().further_than(block);
-	}
-
-	fn try_next_peer(&mut self) -> Option<PeerId> {
+	/// Returns the next peer from the peers cache, if available.
+	///
+	/// If the optional block number if provided, the peers cache is
+	/// repopulated from the live peers that have voted on higher block numbers.
+	fn try_next_peer(&mut self, reset_cache: Option<NumberFor<B>>) -> Option<PeerId> {
```

Comment:
This extra `Option` on this fn makes the code harder to read and reason about.

Is it making any difference in practice, at least? From what I see, all it achieves is locking once for longer rather than lock, release, lock.

Is this improving the behavior or lowering lock thrashing?

Reply:
Indeed, don't think the double-lock is having a huge impact on performance :) Just noticed it while digging for some metrics

When we lock release lock, there might be a chance that the peers we have in cache will disconnect in between the locks (and also get propagated to live peers). But I think we have enough info with the metrics and extra logs to catch these.

TLDR; lets keep the code simple :)